Ensure that semantic search entrypoints are safe to call without init!

appleby opened the pull request:
Closes BOT-265

### Description

When booting a new install from a fresh app db, we can wind up getting calls into the search backend from things like loading the sample or audit db content before search has been initialized, and therefore before receiving an `init!` call. The same is true for the appdb backend, but the appdb backend is resilient to such calls and either ensures the index is created when needed (e.g. in `reindex!`) or treats the call as a no-op if no index has been created (e.g. `delete!`).

Ideally calls into search would not happen until after init, and we should probably fix this bug, but it seems sensible to ensure we don't throw in the semantic search backend if it does happen.

* Ensure that semantic search entrypoints are safe to call without `init!`
* Move more helpers into `semantic.test-util namespace`
* Add tests for calling semantic search entrypoints prior to calling `init!`

### How to verify

Wipe your appdb and restart metabase. Prior to these changes, you will see a handful of errors like so followed by a stacktrace. These come from search attempting to update the index when loading the sample db and/or audit db content. After these changes, no errors should appear.

```
2025-07-25 00:40:07,037 ERROR semantic.core :: Error updating semantic search index 
org.postgresql.util.PSQLException: ERROR: relation "index_table__ollama__mxbai-embed-large__1024" does not exist
```
### Checklist

- [x] Tests have been added/updated to cover changes in this PR

appleby committed:
Ensure that semantic search entrypoints are safe to call without init!

appleby committed:
Move more helpers into semantic.test-util namespace

appleby committed:
Add tests for calling semantic search entrypoints prior to calling init!

wotbrew commented:
Agreed this is a worthy temporary fix. As mentioned I would copy the fix over to https://github.com/metabase/metabase/pull/61379 though, which will require some work.

Long term, I do not like the loosey-goosey behaviour and would prefer an explicit error or similar signal to callers that you need to wait. 

Either the contract is that search needs to 'work' even if `init!` is not called I think we need to do some design work to define what 'working' means, like is garbage acceptable? Are there security or ranking quality concerns if we hand off a half baked index.

appleby committed:
Adjust fix and tests to work better with new index-metadata apis

And also log a warning and return nil if anyone calls into update-index! or delete-from-index! prior to calling init!
